sdk/messaging/azservicebus/receiver.go

```diff
@@ -196,120 +200,118 @@ func (r *Receiver) ReceiveMessages(ctx context.Context, numMessages int, options
 		}
 	}
 
-	receiver, err := r.updateReceiver(ctx)
+	_, receiver, _, _, err := r.amqpLinks.Get(ctx)
 
 	if err != nil {
 		return nil, err
 	}
 
-	if err := receiver.IssueCredit(uint32(numMessages)); err != nil {
+	if err := receiver.IssueCredit(uint32(maxMessages)); err != nil {
 		return nil, err
 	}
 
 	var messages []*ReceivedMessage
-	var allMessagesReceivedError = errors.New("all messages received")
 
-	receiveCh, afterFirstMessageFn := startReceiverTimer(ropts.maxWaitTime, ropts.maxWaitTimeAfterFirstMessage)
-
-	listenCtx, cancelListenCtx := context.WithCancel(ctx)
-	defer cancelListenCtx()
+	ctx, cancel := context.WithTimeout(ctx, ropts.maxWaitTime)
+	defer cancel()
+
+	// the last "interesting" error to occur in the sequence. Used only if we end up receiving zero
+	// messages (otherwise, the user won't even know to look at the return value).
+	var fatalError error
+
+	// receive until we get our first interruption:
+	// - context is cancelled because the user cancelled it
+	// - context is cancelled because one of our timeouts hit
+	// - we get all the messages we asked for
+	for {
+		// TODO: should we do retries here? Or should we just surface the error close to immediately?
+		amqpMessage, err := receiver.Receive(ctx)
+
+		if err != nil {
+			// at the moment all errors from Receive() are fatal (if they're not context cancellation errors)
+			// for simplicity, we'll just store whatever we got and return it later if nothing else goes bad.
+			fatalError = err
+
+			if err != context.Canceled && err != context.DeadlineExceeded {
```

Comment:
Use `errors.Is()` for sentinel error checks.

Reply:
Done.

I made an isCancelled() function and just went and replaced all the spots where I was actually doing this incorrectly.